Inert attribute should make text not findable by find-in-page

sideshowbarker opened the pull request:
#### 4ade1283453851f148bae117c4ee4a7d08c24bb2
<pre>
Inert attribute should make text not findable by find-in-page
<a href="https://bugs.webkit.org/show_bug.cgi?id=269909">https://bugs.webkit.org/show_bug.cgi?id=269909</a>

Reviewed by NOBODY (OOPS!).

This change makes WebKit conform to the HTML spec requirement at
<a href="https://html.spec.whatwg.org/multipage/interaction.html#inert-subtrees">https://html.spec.whatwg.org/multipage/interaction.html#inert-subtrees</a>,
which states that if a node is inert, “the user agent should ignore the
node for the purposes of find-in-page&quot;, and gets WebKit passing the WPT
test at <a href="https://wpt.fyi/results/inert/inert-and-find.html">https://wpt.fyi/results/inert/inert-and-find.html</a>

Otherwise, without this change, WebKit doesn’t conform to the “ignore
inert nodes for the purposes of find-in-page” spec requirement, and
doesn’t pass the WPT test which checks conformance to that requirement.

* LayoutTests/imported/w3c/web-platform-tests/inert/inert-and-find-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/inert/inert-and-find.html: Added.
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::TextIterator::advance):
* Source/WebCore/editing/TextIterator.h:
* Source/WebCore/editing/TextIteratorBehavior.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4ade1283453851f148bae117c4ee4a7d08c24bb2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/163479 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/36962 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/30178 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/172419 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/119926 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/165348 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/37291 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/36962 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/126967 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/89501 "layout-tests (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/166438 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/29239 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/146970 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/107525 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/28288 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/26916 "Passed tests") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/20121 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/138185 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/24688 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/174923 "Built successfully") | 
| | [✅ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/27858 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/26351 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/135276 "Passed tests") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/36632 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/31021 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/135256 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/37417 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/146483 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/99423 "Built successfully") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/30556 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/23334 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/36128 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/109245 "Build is in progress. Recent messages:OS: Tahoe (26.3.1), Xcode: 26.2; Checked out pull request; Found 32423 issues") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/35625 "Built successfully") | [✅ 🧪 mac-site-isolation](https://ews-build.webkit.org/#/builders/185/builds/1355 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/35873 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/35773 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d15d34db225a1fd8fa81fd2f0612d65b7457d0b2. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/25161

rniwa commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -3692,7 +3692,7 @@ bool Editor::findString(const String& target, FindOptions options)
     }
 
-    if (!resultRange)
+    if (!resultRange || resultRange->startContainer().renderer()->style().effectiveInert())
```
This isn't quite right. `rangeOfString` needs to be skipping inert contents instead of making this function return false at the last minute like this. Specifically, we probably need to make `TextIterator` skip inert content. See how it checks `renderer->isSkippedContentRoot()` to skip contents for `content-visibility: hidden`.

sideshowbarker replied:
OK, a lot of test failures already at https://ews-build.webkit.org/#/builders/1/builds/40991 and https://ews-build.webkit.org/#/builders/36/builds/37244 — and looking at them and thinking about it further makes realize: We don’t actually want/need to be changing `TextIterator::advance` to be ignoring (skipping over) inert nodes in all cases, right?

I mean, that’s not what the spec defines. The relevant spec requirement here is instead actually limited just to the case of find-in-page and ignoring inert nodes in just that case.

So doesn’t the behavior for that specific find-in-page requirement actually need to be implemented in `Editor::findString` after all?

sideshowbarker replied:
> `rangeOfString` needs to be skipping inert contents instead of making this function return false at the last minute like this.

OK, after stepping back and taking some time to just think about it some more, I realize that maybe what you were getting at was something more like, passing a new _option_ to `rangeOfString` that would make it optionally skip inert nodes in the cases where that’s what the calling code needs. (Which right now is just this find-in-page case, but it’s imaginable that more cases could come along later.)

Anyway, I’ll take shot at reworking the patch along those lines.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f7e87138b80ce6b793056681170f0c60af835440. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/25161

rniwa commented on Source/WebCore/editing/TextIterator.cpp:
```diff
@@ -478,7 +478,7 @@ void TextIterator::advance()
                 m_handledNode = true;
                 m_handledChildren = !hasDisplayContents(*protectedCurrentNode()) && !renderer;
-            } else if (is<Element>(m_currentNode.get()) && renderer->isSkippedContentRoot()) {
+            } else if (is<Element>(m_currentNode.get()) && (renderer->isSkippedContentRoot() || renderer->style().effectiveInert())) {
```
Are we sure this won't affect [`element.innerText`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/innerText)?

sideshowbarker replied:
I hadn’t considered what would happen with [element.innerText](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/innerText) but I’ll test it and see.

By the way: given that per-spec, nodes (not just elements) can be inert, I wonder if we need to have this `renderer->style().effectiveInert()))` check performed on its own on as a separate condition — rather than in combination with checking whether `m_currentNode` is an element node.

And I guess we ideally should have separate test for the inert-node-that’s-not-an-element-node case, if possible. 

It’s not actually immediately clear to me how a node that’s not a element _could_ in fact also be inert — I guess I need to read up on the spec and see (and so I can write up a test for it).

sideshowbarker replied:
Or actually (just thinking out loud) maybe it’s the case that nodes other than element nodes are anyway not going to be exposed to find-in-page (and so wouldn’t matter for the purposes of this patch). Anyway, I’ll read through the spec for inert right now

sideshowbarker replied:
OK I see now that per-spec at https://html.spec.whatwg.org/multipage/interaction.html#modal-dialogs-and-inert-subtrees it’s the case that text nodes can become inert due to being outside an open modal dialog. So we should probably get a test added for that case. Actually, it wouldn’t just be for text nodes outside an open modal dialog, but for element nodes too — because those elements wouldn’t have an explicit `inert` attribute.

I imagine the existing `dialog` implementation already handles making those nodes inert — `effectiveInert()` style — but I’ll look. And I’ll also look at the existing `dialog` tests to see what might be repurposable to use as a starting point for a insert-with-dialog-and-find-in-page test

aproskuryakov commented:
Please note that there is discussion about this change in https://bugs.webkit.org/show_bug.cgi?id=269909. While this aims to improve a legitimate use case, I'm concerned that this will break more than it will fix over time.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6bea65300da9dd0a7f4333199aa96edb95e87a19. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/25161

aprotyas commented on Source/WebKit/Shared/WebFindOptions.h:
```diff
@@ -42,6 +42,7 @@ enum class FindOptions : uint16_t {
     NoIndexChange = 1 << 9,
     AtWordEnds = 1 << 10,
     DoNotSetSelection = 1 << 11,
+    IgnoreInertNodes = 1 << 12,
```
Oh, you also need to teach `WebCore::FindOption core(OptionSet<FindOptions> options)` about this new option.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5eb2dfe16af9d334843bdefe2ac0372487010ab7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/25161

rniwa commented:
The PR looks good from technical stand point. There is a policy open question as to whether we want to merge this change or not.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6f61650fe30c8b4762e4f66b4c84151583cb195b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/25161

sideshowbarker commented:
FYI, I moved setting of the `IgnoreInertNodes` option into the platform code in `UIProcess/mac/WKTextFinderClient.mm` and in `UIProcess/API/glib/WebKitFindController.cpp` — because the OptionSet it gets added to then gets used/shared in multiple functions in `find*` functions in `WebKit::WebPage`.

Otherwise, it would need to be added separately/redundantly in multiple places in `WebKit::WebPage`: in `findStringMatches` and in `findString` — but also in several other `find*` functions in that code. (There are 6 `find*` functions there, into each of which the same `OptionSet` of `FindOptions` gets passed from the platform/UIProcess code.)

pxlcoder commented:
There are a couple more changes necessary for this to work with all uses of find-in-page on Cocoa platforms.

To support iOS and the `UIFindInteraction` API: https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm#L11780

To support all Cocoa platforms and `WKWebView`'s `findString:` API: https://github.com/WebKit/WebKit/blob/4f76c56c5655b26ed84f9f1b1d03a59c87e74885/Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm#L1439

The current change in `WKTextFinderClient.mm` is necessary, but it will only have an effect for users of `NSTextFinder`.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0d5e61333ca23f0b8540d2a975494e63a4582675. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/25161

sideshowbarker commented:
> To support iOS and the `UIFindInteraction` API: [`main`/Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm#L11780](https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm?rgh-link-date=2024-03-01T21%3A32%3A14Z#L11780)
> 
> To support all Cocoa platforms and `WKWebView`'s `findString:` API:
> 
> https://github.com/WebKit/WebKit/blob/4f76c56c5655b26ed84f9f1b1d03a59c87e74885/Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm#L1439

Thanks, pushed an amended commit with those changes — though for the Cocoa case, I made the change in `WebKit/Shared/API/c/WKSharedAPICast.h` rather than in `WebKit/UIProcess/API/Cocoa/WKWebView.mm`.

But if the change really should be in the `WebKit/UIProcess/API/Cocoa/WKWebView.mm` file instead, please let me know, and I’ll figure out how to get it added there.

pxlcoder commented:
> > To support iOS and the `UIFindInteraction` API: [`main`/Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm#L11780](https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm?rgh-link-date=2024-03-01T21%3A32%3A14Z#L11780)
> > To support all Cocoa platforms and `WKWebView`'s `findString:` API:
> > https://github.com/WebKit/WebKit/blob/4f76c56c5655b26ed84f9f1b1d03a59c87e74885/Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm#L1439
> 
> Thanks, pushed an amended commit with those changes — though for the Cocoa case, I made the change in `WebKit/Shared/API/c/WKSharedAPICast.h` rather than in `WebKit/UIProcess/API/Cocoa/WKWebView.mm`.
> 
> But if the change really should be in the `WebKit/UIProcess/API/Cocoa/WKWebView.mm` file instead, please let me know, and I’ll figure out how to get it added there.

That won't actually have an effect for `WKWebView`, since that uses the `inline OptionSet<WebKit::FindOptions> toFindOptions(WKFindConfiguration *configuration)`. So you'd need to update that method specifically.

Thinking about this a bit more, and making the assumption that the open policy question is resolved, is there a reason you didn't always add the new option in `findIteratorOptions`? An alternative would be to introduce something like `DoesNotIgnoreInertNodes`.

sideshowbarker commented:
> Thinking about this a bit more, and making the assumption that the open policy question is resolved, is there a reason you didn't always add the new option in `findIteratorOptions`?

I will try adding it `findIteratorOptions` and see — but when in https://github.com/WebKit/WebKit/commit/f7e87138b80ce6b793056681170f0c60af835440#diff-fcd6c1d0445582c86908417397e4355bfc5ae768f3f29f987d6f113a88a0c80a I had first tried adding it directly to `TextIterator.cpp`, that regressed existing tests in [ews-build.webkit.org/#/builders/1/builds/40991](https://ews-build.webkit.org/#/builders/1/builds/40991) and [ews-build.webkit.org/#/builders/36/builds/37244](https://ews-build.webkit.org/#/builders/36/builds/37244) — due, I assumed to there being existing behavior using  `TextIterator::advance` other than find-in-page that relied on inert nodes not getting skipped.

But it certainly would be simpler if adding it to `findIteratorOptions` worked as expected without regressing any other behavior/tests — so I will go ahead and give it a try.

sideshowbarker committed:
Inert attribute should make text not findable by find-in-page
https://bugs.webkit.org/show_bug.cgi?id=269909

Reviewed by NOBODY (OOPS!).

This change makes WebKit conform to the HTML spec requirement at
https://html.spec.whatwg.org/multipage/interaction.html#inert-subtrees,
which states that if a node is inert, “the user agent should ignore the
node for the purposes of find-in-page", and gets WebKit passing the WPT
test at https://wpt.fyi/results/inert/inert-and-find.html

Otherwise, without this change, WebKit doesn’t conform to the “ignore
inert nodes for the purposes of find-in-page” spec requirement, and
doesn’t pass the WPT test which checks conformance to that requirement.

* LayoutTests/imported/w3c/web-platform-tests/inert/inert-and-find-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/inert/inert-and-find.html: Added.
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::TextIterator::advance):
* Source/WebCore/editing/TextIterator.h:
* Source/WebCore/editing/TextIteratorBehavior.h:

sideshowbarker committed:
Inert attribute should make text not findable by find-in-page
https://bugs.webkit.org/show_bug.cgi?id=269909

Reviewed by NOBODY (OOPS!).

This change makes WebKit conform to the HTML spec requirement at
https://html.spec.whatwg.org/multipage/interaction.html#inert-subtrees,
which states that if a node is inert, “the user agent should ignore the
node for the purposes of find-in-page", and gets WebKit passing the WPT
test at https://wpt.fyi/results/inert/inert-and-find.html

Otherwise, without this change, WebKit doesn’t conform to the “ignore
inert nodes for the purposes of find-in-page” spec requirement, and
doesn’t pass the WPT test which checks conformance to that requirement.

* LayoutTests/imported/w3c/web-platform-tests/inert/inert-and-find-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/inert/inert-and-find.html: Added.
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::TextIterator::advance):
* Source/WebCore/editing/TextIterator.h:
* Source/WebCore/editing/TextIteratorBehavior.h:

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/daacf1161fc85b5c8f4cd98f5cddef307938f10e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/25161

sideshowbarker commented:
> > Thinking about this a bit more, and making the assumption that the open policy question is resolved, is there a reason you didn't always add the new option in `findIteratorOptions`?
> 
> I will try adding it `findIteratorOptions`

Thanks much for bringing it up — that works, without any regressions.

In hindsight, it seems completely obvious that’s what was needed. I guess I just had my head down so far that I neglected to see it clearly

> I had first tried adding it directly to `TextIterator.cpp`, that regressed existing tests in [ews-build.webkit.org/#/builders/1/builds/40991](https://ews-build.webkit.org/#/builders/1/builds/40991) and [ews-build.webkit.org/#/builders/36/builds/37244](https://ews-build.webkit.org/#/builders/36/builds/37244) — due, I assumed, to there being existing behavior using `TextIterator::advance` other than find-in-page that relied on inert nodes not getting skipped.

Of course that was because I’d initially just unconditionally made `TextIterator::advance` ignore inert nodes.

Anyway, thanks again very much. Very glad to avoid having to touch multiple places in the UIProcess code — and to avoid need to even touch `FindOptions` at all. The patch is now just a 3-line code change.

aproskuryakov reviewed (changes requested):
Marking as "request changes" to avoid confusion, as the discussion in Bugzilla has stalled.

I don't think that I'm the decision maker for this, but I do think that these comments need to be addressed substantively.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f709fef84e75c5e731edaea9d1f5ab1d545dcb79. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/25161

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b93c3c38363b8c192897203151c0ac3b476d694e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/25161

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/32e099ff08220479473c0bfba6ca139182d00a3c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/25161
